common/persistence/sql/sqlplugin/db_handle.go

```diff
@@ -0,0 +1,197 @@
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package sqlplugin
+
+import (
+	"context"
+	"database/sql"
+	"database/sql/driver"
+	"errors"
+	"fmt"
+	"io"
+	"sync"
+	"syscall"
+	"time"
+
+	"github.com/jmoiron/sqlx"
+	uberatomic "go.uber.org/atomic"
+
+	"go.temporal.io/api/serviceerror"
+	"go.temporal.io/server/common/log"
+	"go.temporal.io/server/common/log/tag"
+	"go.temporal.io/server/common/metrics"
+)
+
+const (
+	// TODO: this should be dynamic config. For now we reuse the same setting as our cassandra implementation
+	sessionRefreshMinInternal = 5 * time.Second
+)
+
+var (
+	DatabaseUnavailableError = serviceerror.NewUnavailable("no usable database connection found")
+)
+
+type DatabaseHandle struct {
+	running      bool
+	db           uberatomic.Pointer[sqlx.DB]
+	connect      func() (*sqlx.DB, error)
+	needsRefresh func(error) bool
+
+	lastRefresh time.Time
+	metrics     metrics.Handler
+	logger      log.Logger
+	// Ensures only one refresh call happens at a time
+	sync.Mutex
+}
+
+// An invalid connection returns `databaseUnavailable` for all operations
+type invalidConn struct{}
+
+func NewDatabaseHandle(
+	connect func() (*sqlx.DB, error),
+	needsRefresh func(error) bool,
+	logger log.Logger,
+	metricsHandler metrics.Handler,
+) *DatabaseHandle {
+	handle := &DatabaseHandle{
+		running:      true,
+		connect:      connect,
+		needsRefresh: needsRefresh,
+		metrics:      metricsHandler,
+		logger:       logger,
+	}
+	handle.reconnect(true)
+	return handle
+}
+
+// Close and reopen the underlying database connection
+func (h *DatabaseHandle) reconnect(force bool) {
+	h.Lock()
+	defer h.Unlock()
+
+	// Don't reconnect if we've been closed
+	if !h.running {
+		return
+	}
+
+	if !force && h.db.Load() != nil {
+		// Another goroutine already reconnected
+		return
+	} else if prevConn := h.db.Swap(nil); prevConn != nil {
+		// Store `nil` to prevent other goroutines from slamming the now-unusable database with
+		// transactions we know will fail
+		go prevConn.Close()
+	}
+
+	metrics.PersistenceSessionRefreshAttempts.With(h.metrics).Record(1)
+
+	now := time.Now()
+	lastRefresh := h.lastRefresh
+	h.lastRefresh = now
+	if now.Sub(lastRefresh) < sessionRefreshMinInternal {
+		h.logger.Warn("sql handle: did not refresh database connection pool because the last refresh was too close",
+			tag.NewDurationTag("min_refresh_interval_seconds", sessionRefreshMinInternal))
+		handler := h.metrics.WithTags(metrics.FailureTag("throttle"))
+		metrics.PersistenceSessionRefreshFailures.With(handler).Record(1)
+		return
+	}
+
+	newConn, err := h.connect()
+	if err != nil {
+		h.logger.Error("sql handle: unable to refresh database connection pool", tag.Error(err))
+		handler := h.metrics.WithTags(metrics.FailureTag("error"))
+		metrics.PersistenceSessionRefreshFailures.With(handler).Record(1)
```

Comment:
cass also has a refresh latency metrics I remember. not sure if it's useful here.